### DISCUSSION TITLE
JS-IPFS failure to bundle with Vite(also webpack but Vite is the goal)

### DISCUSSION THREAD

**Discussion by AustinFoss** [nuxt/nuxt#17424] [Questions]
### Environment

- Operating System: `Linux`
- Node Version:     `v16.11.1`
- Nuxt Version:     `3.0.0-rc.3`
- Package Manager:  `yarn@1.22.18`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

`npx nuxi init nuxt-app \`
`cd nuxt-app \`
`yarn install \`
`yarn add ipfs-core \`
`mkdir pages plugins \`

- `nuxt-app/app.vue`; replace NuxtWelcome with NuxtPage
- `nuxt-app/pages/index.vue`; This is just here to enable the router(the following plugins file wouldn't work without router enabled). Just a vanilla template, no script, maybe a p tag or something
- `nuxt-app/plugins/ipfs.ts`; looks like this

`

    import {create} from 'ipfs-core'

    export default defineNuxtPlugin(() => {
    return {
      provide: {
        ipfs: () => create()
      }
    }
    })
`

`yarn dev -o`

### Describe the bug

Without using the plugin anywhere, just importing the module and creating the js-ipfs node triggers a weird loop. The following happens if implemented with pinia instead of using a plugin as well. To avoid the loop and just see the single error skip the `-o` in `yarn dev`.

Output:

✘ [ERROR] Could not resolve "electron"

    node_modules/electron-fetch/lib/index.es.js:1257:21:
      1257 │   electron = require('electron');
           ╵                      ~~~~~~~~~~

  You can mark the path "electron" as external to exclude it from the bundle, which will remove this
  error. You can also surround this "require" call with a try/catch block to handle this failure at
  run-time instead of bundle-time.

 ERROR  4:40:05 PM [vite] error while updating dependencies:                                                                    16:40:05
Error: Build failed with 1 error:
node_modules/electron-fetch/lib/index.es.js:1257:21: ERROR: Could not resolve "electron"
    at failureErrorWithLog (/home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:1603:15)
    at /home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:1249:28
    at runOnEndCallbacks (/home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:1034:63)
    at buildResponseToResult (/home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:1247:7)
    at /home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:1356:14
    at /home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:666:9
    at handleIncomingPacket (/home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:763:9)
    at Socket.readFromStdout (/home/foss/gits/nuxt-app/node_modules/esbuild/lib/main.js:632:7)
    at Socket.emit (node:events:390:28)
    at Socket.emit (node:domain:475:12)

 ERROR  Vite Error, /_nuxt/node_modules/.vite/deps/ipfs-core.js?v=0a934655 optimized info should be defined                     16:40:05            

If I try doing `yarn add electron` (which I don't think I need) the following happens with more messages saying"optimized info should be defined" for other deps;

✘ [ERROR] No matching export in "browser-external:url" for import "parse"

    node_modules/electron-fetch/lib/index.es.js:1:9:
      1 │ import { parse, format, resolve } from 'url';
        ╵          ~~~~~

✘ [ERROR] No matching export in "browser-external:url" for import "format"

    node_modules/electron-fetch/lib/index.es.js:1:16:
      1 │ import { parse, format, resolve } from 'url';
        ╵                 ~~~~~~

✘ [ERROR] No matching export in "browser-external:url" for import "resolve"

    node_modules/electron-fetch/lib/index.es.js:1:24:
      1 │ import { parse, format, resolve } from 'url';
        ╵                         ~~~~~~~

✘ [ERROR] No matching export in "browser-external:http" for import "STATUS_CODES"

    node_modules/electron-fetch/lib/index.es.js:3:9:
      3 │ import { STATUS_CODES } from 'http';
        ╵          ~~~~~~~~~~~~

✘ [ERROR] No matching export in "browser-external:stream" for import "PassThrough"

    node_modules/electron-fetch/lib/index.es.js:6:17:
      6 │ import Stream, { PassThrough } from 'stream';
        ╵                  ~~~~~~~~~~~

### Additional context

I've also tried disabling Vite for Webpack5 with no luck, that produces a different set of errors, besides I'd like to get this working in Vite.

I've taken a look at this repository but also had no luck; https://github.com/DougAnderson444/ipfs-vite-svelte-kit

Can provide more info on why that failed if requested. I thought Vite used esbuild under the hood anyway so I'm not sure why that repo was even needed to adapt ipfs-core for vite.

I've previously built projects with Nuxt2 using ipfs-core in the 2020-2021 without issue. Pretty sure this because I'm using all fairly new tooling and RC builds of those tools, but if anyone has any insight into why this may be happening I would greatly appreciate it.

**Comment by HomWang** [comment]
1.rm -rf node_modules
2.rm -rf ***.lock(yarn.lock)
3.npm i -D electron@latest
4.Does the package used support vue3 and vite

I hope I can help you

**Reply by AustinFoss** [reply]
Attempting steps 1-3 unfortunately. After adding `electron`, it then gives me similar prompts for `url http stream` modules. When I do the same for those 3 it gives this error:

`
[plugin vite:dep-pre-bundle] Failed to resolve entry for package "http". The package may have incorrect main/module/exports specified in its package.json: Failed to resolve entry for package "http". The package may have incorrect main/module/exports specified in its package.json.
`

Will continue this thought in my response Danielroe below.

**Reply by AustinFoss** [reply]
Step 4 I attempted a vite project with Vue3(no Nuxt), and had the same behavior. When I attempted just a Vue3 project I got the Webpack error mentioned above when I disabling Vite in Nuxt3.

Webpack5 compatibility is a known issue for the js-ipfs module, see here: https://github.com/ipfs/js-ipfs/issues/3369

But rather than put the same amount of extra effort into getting the older, slower, builder to work I'd like to put that effort towards getting it to work with Vite if possible. I'm way early and I could maybe just be more patient until it's made fully compatible with Vite by the maintainers. 

I thought I was using Vue 3 last year when it was working, but now that I'm thinking more about it I probably was using Vue2(hence older webpack) and just importing the composition API and not true Vue3 or Webpack5 at any point.

**Comment by danielroe** [comment]
This sounds like an issue of ESM support, as it's not valid to use a dynamic `require` call within an ES module.

https://v3.nuxtjs.org/concepts/esm

You may be able to resolve some of this by adding these packages to your `build.transpile` array. For more help, I'd suggest providing a reproduction.

I've converted to a discussion as I think it's not a Nuxt issue but would still benefit from help. If it ends up looking like a bug, we can always raise it as an issue again.

**Reply by HomWang** [reply]
Yes, this is not a problem with nuxt. It is caused by the lack of support for third-party packages

**Reply by AustinFoss** [reply]
Unfortunately for some reason adding
`
build:
    transpile: ['ipfs-core']
`
To my config doesn't work. No error in the terminal, but the browser console shows `Uncaught SyntaxError: Unexpected string` which is weird because all that's supposed to be there is a string array.

I agree this is ESM support related. This has been the biggest deep dive into builder(webpack & vite) I've done so far while trying to figure it out. Quite a lot to sift through.

Discussion was a good idea, I'm still getting used to github best practices.

As far as a reproduction goes isn't that what I did at the top? Or are you asking for a direct github repo to clone? I thought because it was just the base project, just the standard boilerplate Nuxt Generates from RC-3 with the ipfs-core module attempting to be imported providing a project repo would be over kill. Can definitely do so if you would like to continue investigating this with me.

On that note though, I'm wondering if it would be more appropriate for me to move this to the Vite github directly instead of the Nuxt Repo?

What would be your opinion on this work around? https://github.com/DougAnderson444/ipfs-vite-svelte-kit

I attempted it without luck but maybe I need to do more reading into esbuild and tweak the build:ipfs script. Not too sure what the problem was. For that I can definitely provide a complete Nuxt repo with that included.

**Comment by trenta3** [comment]
So I'm having the same problem, I could progress a little bit by installing `npm i -D electron http-browserify stream-browserify url rollup-plugin-node-polyfills` and using this `svelte.config.js`:
```js
import adapter from '@sveltejs/adapter-auto'
import polyfills from 'rollup-plugin-node-polyfills'

/** @type {import('@sveltejs/kit').Config} */
const config = {
    kit: {
	adapter: adapter(),
	vite: {
	    resolve: {
		alias: {
		    "stream": "stream-browserify",
		    "http": "http-browserify",
		    "url": "url",
		},
	    },
	    build: {
		target: ["es2020"],
	    },
	    plugins: [
		polyfills({ crypto: true }),
	    ],
	},
    }
};

export default config;
```

Unfortunately after doing all this what I get from SvelteKit is error 500 "missing name after . operator", which is quite unsatisfactory but at least I solved a couple of problems I was having with imports. Maybe someone will be able to progress further.
